base/abstractarray.jl

```diff
@@ -1962,6 +1981,218 @@ function typed_hvcat(::Type{T}, rows::Tuple{Vararg{Int}}, as...) where T
     T[rs...;]
 end
 
+# nd concatenation
+
+"""
+    hvncat(dims::Tuple{Vararg{Int}}, row_first::Bool, values...)
+
+Horizontal, vertical, and n-dimensional concatenation in one call. This function is called
+for block matrix syntax. The first argument specifies the number of arguments in `values`
+along the first axis for each dimension. `row_first` indicates whether `values` are ordered
+first by rows (true) or first by columns (false).
+
+# Examples
+```jldoctest
+julia> a, b, c, d, e, f = 1, 2, 3, 4, 5, 6
+(1, 2, 3, 4, 5, 6)
+
+julia> [a b c;;; d e f]
+1×3×2 Array{Int64, 3}:
+[:, :, 1] =
+ 1  2  3
+
+[:, :, 2] =
+ 4  5  6
+
+julia> hvncat((2,1,3), false, a,b,c,d,e,f)
+2×1×3 Array{Int64, 3}:
+[:, :, 1] =
+ 1
+ 2
+
+[:, :, 2] =
+ 3
+ 4
+
+[:, :, 3] =
+ 5
+ 6
+
+julia> [a b;;; c d;;; e f]
+1×2×3 Array{Int64, 3}:
+[:, :, 1] =
+ 1  2
+
+[:, :, 2] =
+ 3  4
+
+[:, :, 3] =
+ 5  6
+```
+"""
+hvncat(::Tuple{}, ::Bool) = []
+hvncat(::Tuple{}, ::Bool, xs...) = []
+hvncat(::Tuple{Int}, ::Bool, xs...) = vcat(xs...) # methods assume 2+ dimensions
+hvncat(dims::Tuple{Vararg{Int}}, row_first::Bool, xs...) = _hvncat(dims, row_first, xs...)
+
+_hvncat(::Tuple{Vararg{Int}}, ::Bool) = []
+_hvncat(dims::Tuple{Vararg{Int}}, row_first::Bool, xs...) = _typed_hvncat(promote_eltypeof(xs...), dims, row_first, xs...)
+_hvncat(dims::Tuple{Vararg{Int}}, row_first::Bool, xs::T...) where T<:Number = _typed_hvncat(T, dims, row_first, xs...)
+_hvncat(dims::Tuple{Vararg{Int}}, row_first::Bool, xs::Number...) = _typed_hvncat(promote_typeof(xs...), dims, row_first, xs...)
+_hvncat(dims::Tuple{Vararg{Int}}, row_first::Bool, xs::AbstractArray...) = _typed_hvncat(promote_eltype(xs...), dims, row_first, xs...)
+_hvncat(dims::Tuple{Vararg{Int}}, row_first::Bool, xs::AbstractArray{T}...) where T = _typed_hvncat(T, dims, row_first, xs...)
+
+typed_hvncat(::Type{T}, ::Tuple{}, ::Bool) where T = Vector{T}()
+typed_hvncat(::Type{T}, ::Tuple{}, ::Bool, xs...) where T = Vector{T}()
+typed_hvncat(T::Type, ::Tuple{Int}, ::Bool, xs...) = typed_vcat(T, xs...) # methods assume 2+ dimensions
+typed_hvncat(T::Type, dims::Tuple{Vararg{Int}}, row_first::Bool, xs...) = _typed_hvncat(T, dims, row_first, xs...)
```

Comment:
Should these guys get the `::Type{T}` treatment, too?

Reply:
I just removed them because I thought we wouldn't want to force them 
 to specialize haha. I noticed several existing hvcat methods that passed the type through did the same.